dali/kernels/common/cast.cuh

```diff
@@ -26,17 +26,48 @@ struct CastSampleDesc {
   const void *input;
 };
 
+struct CastSampleBlockDesc {
+  int first_block;
+  int sample_size;
+};
+
 template <typename OType, typename IType>
-__global__ void BatchedCastKernel(const CastSampleDesc *samples, const BlockDesc<1> *blocks) {
-  const auto &block = blocks[blockIdx.x];
-  const auto &sample = samples[block.sample_idx];
+__device__ __forceinline__ void CastKernelInternal(const CastSampleDesc& sample,
+                                                   int block_start, int block_end) {
   auto *out = static_cast<OType *>(sample.output);
   const auto *in = static_cast<const IType *>(sample.input);
-  for (int x = threadIdx.x + block.start.x; x < block.end.x; x += blockDim.x) {
+  for (int x = threadIdx.x + block_start; x < block_end; x += blockDim.x) {
     out[x] = ConvertSat<OType>(in[x]);
   }
 }
 
+template <typename OType, typename IType>
+__global__ void BatchedCastKernel(const CastSampleDesc *samples, const BlockDesc<1> *blocks) {
+  const auto &block = blocks[blockIdx.x];
+  const auto &sample = samples[block.sample_idx];
+  CastKernelInternal<OType, IType>(sample, block.start.x, block.end.x);
+}
+
+template <typename OType, typename IType>
+__global__ void BinSearchCastKernel(const CastSampleDesc *samples,
+                                    const CastSampleBlockDesc *params,
+                                    int nsamples, int block_volume_scale) {
+  int i = 0;
+  for (int jump = (1 << (32 - __clz(nsamples) - 1)); jump; jump /= 2) {
```

Comment:
How about making `(1 << (32 - __clz(nsamples) - 1))` a kernel parameter? After all, it's a fixed value.
```suggestion
  for (int jump = (1 << (32 - __clz(nsamples) - 1)); jump; jump >>= 1) {
```
`/ 2` and `>> 1` are not the same for `int`.

Reply:
As ```nsamples``` is used in the that kernel in another place it would mean adding 5th parameter. Also it doesn't have a noticeable effect on performance